### ISSUE TITLE
Support different visibility levels for hidden implementation other than private

### ISSUE THREAD

**Issue by rtedin** [immutables/immutables#211]
First of all thanks for the great work!

We do not want to expose the immutable implementation to our clients. Hence we are currently using `Value.Style.ImplementationVisibility.PRIVATE` on an interface `A` so that the implementation is hidden inside the generated `ABuilder`. This causes nevertheless a problem when deserializing `A` since Jackson needs to know the concrete type to instantiate.

The problem is that we don't have access to the class of the hidden immutable implementation. We would like to annotate `A` with `@JsonDeserialize(as=ABuilder.ImmutableA.class)` or map `A.class` to `ABuilder.ImmutableA.class` otherwise, but for that we need access to `ABuilder.ImmutableA.class`.

This could be achieved supporting different visibility levels of the immutable inner class other than private. For our case, package would work the best as this would make `ABuilder.ImmutableA` accessible only from the same package.

A workaround is to, at run time, create a dummy instance of `ImmutableA` with the `ABuilder`, get the class with `getClass()` and do the mapping from the abstract type to the implementation. But this gets very cumbersome for classes that encapsulate other immutable classes that suffer from the same problem. Moreover, depending on your code base, there might be no place where to do this at run time.

**Comment by elucash** [comment]
Thank you for the issue report! Of course this should be solved somehow without nasty workarounds. Have you tried using `Value.Style.ImplementationVisibility.PACKAGE`? Is there some downsides of using public abstract value types and generated top-level package-private immutable classes? Am I missing some important detail?

**Comment by rtedin** [comment]
Let's say we have an interface `A`:

```java
@Value.Immutable
@Value.Style(visibility = Value.Style.ImplementationVisibility.PRIVATE)
public interface A {
}
```

This is the generated builder:

```java
@SuppressWarnings("all")
@Generated({"Immutables.generator", "A"})
public final class ABuilder {

  // ...

  /**
   * Immutable implementation of {@link A}.
   * <p>
   * Use builder to create immutable instances:
   * {@code new ABuilder()}.
   */
  private static final class ImmutableA implements A {
    // ... 
  }
}
```

We like it! We expose only the interface `A` and the builder to our clients. We see `ImmutableA` as an implementation detail. The clients never use `ImmutableA` in their code. We think that this is useful if the client wants to extend the functionality of `A`. Being an interface, it gives more freedom than an abstract class. This also favors composition instead of inheritance. It gives us also more freedom if we decide to refactor. The problem is that `ImmutableA` has private access. We can not use `Immutable.class` directly in annotations.

If we use `Value.Style.ImplementationVisibility.PACKAGE`, this is generated instead:

```java
@SuppressWarnings("all")
@Generated({"Immutables.generator", "A"})
final class ImmutableA implements A {
  // ...
  public static final class Builder {
    // ...
  }
}
```

Now the builder is inside the `ImmutableA` and both are exposed to the client. Well actually they aren't: `ImmutableA` is package private. Now the clients can't even create instances if their code is in an other package. Hence we can't use this visibility at all.

**Comment by rtedin** [comment]
We think that it would be very nice if we had two basic choices for configuration:

* Immutable class as an inner class of the builder (we prefer this one, but others might not).
* Builder as an inner class of the immutable class.

And then for each of them being able to chose which visibility we need for each component, independent of the visibility of the interface or abstract class annotated with `@Value.Immutable`. This would give all the freedom we need. This might require bigger changes though (may be new styles?).

**Comment by rtedin** [comment]
As an example, this is how we would have solved our problem if we had all the configuration choices mentioned above:

```java
@Value.Immutable
@Value.Style(
        innerBuilder = false, // default true
        builderVisibility = Value.Style.ImplementationVisibility.SAME,
        visibility = Value.Style.ImplementationVisibility.PACKAGE
)
public interface A {
}
```

The above would generate a builder with an immutable class as an inner class. The builder's visibility would be the same as the one of `A` (here public). The immutable's visibility would be package private.

**Comment by elucash** [comment]
I got what you need, what you described is more or less in line how it should work, i.e. we need more orthogonality in configuration. It might be tricky to preserve backward compatibility. The good part is that internally we have all this made possible, it is the configuration that should better express the intent.

**Comment by rtedin** [comment]
Great! Looking forward to it. Thanks!

**Comment by rtedin** [comment]
Thank you for the quick implementation!

I'm playing a bit with it and found that the following configuration results in a compilation error:

```java
@Value.Immutable
@Value.Style(
        implementationNestedInBuilder = true,
        builderVisibility = Value.Style.BuilderVisibility.PACKAGE
)
public interface AInterface {
    String getString();
}
```

The error appears also if the interface if package private. The error is:

```
.../AInterfaceBuilder.java:[163,50] error: cannot find symbol
```

`ImmutableAInterface` does not have a method `build()`:

```java
      return AInterfaceBuilder.ImmutableAInterface.builder()
          .from(instance)
          .build();
```

I think this line should instead be:

```java
      return AInterfaceBuilder.builder()
          .from(instance)
          .build();
```
Note that the javadoc at line AInterfaceBuilder.java:85 is also wrong.

But the use case we had works! Thanks!

**Comment by elucash** [comment]
yeah, not all combinations are covered, but I'll try to add more tests/fixes

**Comment by elucash** [comment]
Just released in 2.1.1. Thank you!

**Comment by rtedin** [comment]
Thanks, that was quick! However I can not use this yet due to #217.
